Frameworks/CoreText/DWriteWrapper.mm

```diff
@@ -0,0 +1,392 @@
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#include <COMIncludes.h>
+#import <wrl/implements.h>
+#include <COMIncludes_End.h>
+#import <mutex>
+#import <vector>
+#import <algorithm>
+#import <CoreFoundation/CFBase.h>
+#import "CoreTextInternal.h"
+#import "DWriteWrapper.h"
+
+using namespace std;
+using namespace Microsoft::WRL;
+
+static const wchar_t* TAG = L"_DWriteWrapper";
+static const wchar_t* c_defaultUserLanguage = L"en-us";
+static const wchar_t* c_defaultFontName = L"Gabriola";
+
+// TODO::
+// These should be removed once CTFont and UIFont are bridged.
+static IWLazyClassLookup s_lazyUIFont("UIFont");
+static const float c_defaultSystemFontSize = 15.0f;
+
+static ComPtr<IDWriteFactory> CreateDWriteFactoryInstance() {
+    ComPtr<IDWriteFactory> dwriteFactory;
+    THROW_IF_FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), &dwriteFactory));
+    return dwriteFactory;
+}
+
+static ComPtr<IDWriteFactory> GetDWriteFactoryInstance() {
+    static ComPtr<IDWriteFactory> s_dwriteFactory = CreateDWriteFactoryInstance();
+    return s_dwriteFactory;
+}
+
+/**
+ * Helper method to return the user set default locale string.
+ *
+ * @return use set locale string as wstring.
+ */
+static wstring _getUserDefaultLocaleName() {
+    wchar_t localeName[LOCALE_NAME_MAX_LENGTH];
+    int defaultLocaleSuccess = GetUserDefaultLocaleName(localeName, LOCALE_NAME_MAX_LENGTH);
+
+    // If the default locale is returned, find that locale name, otherwise use "en-us".
+    if (defaultLocaleSuccess) {
+        return wstring(localeName);
+    } else {
+        return wstring(c_defaultUserLanguage);
+    }
+}
+
+/**
+ * Helper method to convert IDWriteLocalizedStrings object to NSString object.
+ *
+ * @parameter localizedString IDWriteLocalizedStrings object to convert.
+ *
+ * @return NSString object.
+ */
+static NSString* _convertLocalizedStringToNSString(IDWriteLocalizedStrings* localizedString) {
+    if (localizedString == NULL) {
+        TraceError(TAG, L"The input parameter is invalid!");
+        return nil;
+    }
+
+    // Get the default locale for this user.
+    wstring localeName = _getUserDefaultLocaleName();
+
+    uint32_t index = 0;
+    BOOL exists = false;
+
+    // If the default locale is returned, find that locale name, otherwise use "en-us".
+    THROW_IF_FAILED(localizedString->FindLocaleName(localeName.c_str(), &index, &exists));
+    if (!exists) {
+        THROW_IF_FAILED(localizedString->FindLocaleName(c_defaultUserLanguage, &index, &exists));
+    }
+
+    // If the specified locale doesn't exist, select the first on the list.
+    if (!exists) {
+        index = 0;
+    }
+
+    // Get the string length.
+    uint32_t length = 0;
+    THROW_IF_FAILED(localizedString->GetStringLength(index, &length));
+
+    // Get the string.
+    vector<wchar_t> wcharString = std::vector<wchar_t>(length + 1);
```

Comment:
it does, though. unfortunately, just like `vector`, it will construct the entire range (or here, memset it)

```
template <typename CharT, ...>
basic_string(size_t count, CharT ch, ...);
```

---
In reply to: [77755880](https://github.com/Microsoft/WinObjC/pull/950#discussion_r77755880) [](ancestors = 77755880)